### ISSUE TITLE
Question about 4.1 & 4.2

### ISSUE THREAD

**Issue by wouter406** [TPM034A/Q2_2023#10]
In the question:

_4.1. Benchmark scenario: create a new dataframe containing only trips with a destination in region C (the city centre). Predict the mode shares for these trips, using your trained model. Hint: use the predict_proba function from sk-learn (why should you NOT use the predict function in this case?)_

do you mean the trained model from question 3 or a newly trained model (trained with the dataset containing only trips to region C)? Because if we need to use the trained model from question 3 I'm a bit confused, as in question 4.2:

_4.2. Cordon pricing policy: Create a new dataframe for the cordon pricing. To do so, copy the dataframe created in 4.1. Increase total_car_cost and taxi_cost by £10. Use your trained model to predict the modes. (Remember to scale the data with the scaler created for training the model)._

we need to use the scaler created for training the model. But I already tried using the scaler created in assignment 3 to scale the newly created dataframe from question 4.1 but this doesn't work, because it obviously doesn't contain all the variables the scaler was fit on.

**Comment by spierenbu** [comment]
You indeed need to use the model trained in question 3 for question 4.1 and 4.2. Could you elaborate on your second question? Why the new dataframe does not contain all variables?

**Comment by wouter406** [comment]
I created a new dataset which is a copy of the original RP_mode_choice_data, but with all data entries where the Destination_lsao_name didn't belong to region C filtered out. However when I then dummy encode the remaining Destination_lsao_name column I'm of course left with less columns in the newly created dummy dataframe (as all destinations not in region C aren't dummy encoded and thus not present in the new dummy dataframe). When I then concatanate this with my other modified features and dummy encoded purpose dataframes and try to fit the scaler scaled in q3 I get this error:

ValueError: The feature names should match those that were passed during fit.
Feature names seen at fit time, yet now missing:
- Dropoff Daycare
- K-12 School
- Leeds 001A
- Leeds 001E
- Leeds 002A
- ...

**Comment by spierenbu** [comment]
Thanks for the clarification. Destination_lsao_name should not be included in the features, only the categorical variables destination region and purpose  should be included, in addition to the numerical variables (see question 3.1) .

**Comment by wouter406** [comment]
Ah of course I see! Thx for clarifying!

**Comment by wouter406** [comment]
I actually still have the same issue now. The scaler is trained on a dataframe containing columns for every region (since they're dummy encoded). The new dataframe is missing 5 columns since it only contains data for region C, and thus the original scaler won't fit on the new dataframe. Would a solution be to train the scaler on all numerical features, and add the dummy encoded features after the scaling (as they're inherently already kind of scaled)? Another option I thought of was to delete the last 5 columns (thus the 5 regions to be dropped) from the array returned by scaling the original dataframe. However in this case i'm not creating a new dataframe (as indicated by the question) but just a modified array.

**Comment by spierenbu** [comment]
Yes indeed, the scaler should only include numerical variables. In principle, including dummy variables should not cause too much of an issue in the training process, but in this specific example, you will indeed end up with missing columns.

**Comment by wouter406** [comment]
Ok clear thank you!

**Comment by Michielhs** [comment]
I have a similar problem. The X_training set the model is trained contains more features than the benchmark set, because all the regions are deleted. So when I want to run the model trained on X_training for the benchmark set I got an error that the MLPClassifier is expecting more features

**Comment by spierenbu** [comment]
Such issue may happen if you extract the region before encoding it into dummy variables. You can try to extract the region after the one-hot encoding and the issue should be fixed.

**Comment by spierenbu** [comment]
Yes, you are correct. Since the model is trained on scaled data, you need to feed it scaled data for predictions!
